theseus/geometry/functional/so3.py

```diff
@@ -0,0 +1,158 @@
+    ret[:, 0, 1] -= sine_axis[:, 2]
+    ret[:, 1, 0] += sine_axis[:, 2]
+    ret[:, 0, 2] += sine_axis[:, 1]
+    ret[:, 2, 0] -= sine_axis[:, 1]
+    ret[:, 1, 2] -= sine_axis[:, 0]
+    ret[:, 2, 1] += sine_axis[:, 0]
+
+    return ret
+
+
+def _j_exp_map_impl(tangent_vector: torch.Tensor) -> torch.Tensor:
+    if not check_tangent_vector(tangent_vector):
+        raise ValueError("Tangent vectors of SO3 should be 3-D vectors.")
+    tangent_vector = tangent_vector.view(-1, 3)
+    theta = torch.linalg.norm(tangent_vector, dim=1, keepdim=True).unsqueeze(1)
+    theta2 = theta**2
+    # Compute the approximations when theta ~ 0
+    near_zero = theta < constants._SO3_NEAR_ZERO_EPS[tangent_vector.dtype]
+    non_zero = torch.ones(1, dtype=tangent_vector.dtype, device=tangent_vector.device)
+    theta_nz = torch.where(near_zero, non_zero, theta)
+    theta2_nz = torch.where(near_zero, non_zero, theta2)
+    theta3_nz = theta_nz * theta2_nz
+    sine = theta.sin()
+    theta_minus_sine_by_theta3 = torch.where(
+        near_zero, torch.zeros_like(theta), (theta - sine) / theta3_nz
+    )
+    cosine = torch.where(near_zero, 8 / (4 + theta2) - 1, theta.cos())
+    sine = theta.sin()
+    sine_by_theta = torch.where(near_zero, 0.5 * cosine + 0.5, sine / theta_nz)
+    one_minus_cosie_by_theta2 = torch.where(
+        near_zero, 0.5 * sine_by_theta, (1 - cosine) / theta2_nz
+    )
+
+    jac = (
+        theta_minus_sine_by_theta3
+        * tangent_vector.view(-1, 3, 1)
+        @ tangent_vector.view(-1, 1, 3)
+    )
+    diag_jac = jac.diagonal(dim1=1, dim2=2)
+    diag_jac += sine_by_theta.view(-1, 1)
+
+    jac_temp = one_minus_cosie_by_theta2.view(-1, 1) * tangent_vector
+
+    jac[:, 0, 1] += jac_temp[:, 2]
+    jac[:, 1, 0] -= jac_temp[:, 2]
+    jac[:, 0, 2] -= jac_temp[:, 1]
+    jac[:, 2, 0] += jac_temp[:, 1]
+    jac[:, 1, 2] += jac_temp[:, 0]
+    jac[:, 2, 1] -= jac_temp[:, 0]
+
+    return jac
+
+
+class ExpMap(LieGroup.ExpMap):
+    @classmethod
+    def forward(cls, ctx, tangent_vector):
+        tangent_vector: torch.Tensor = cast(torch.Tensor, tangent_vector)
+        ret = _exp_map_impl(tangent_vector)
+        ctx.save_for_backward(tangent_vector, ret)
+        return ret
+
+    @classmethod
+    def backward(cls, ctx, grad_output):
+        tangent_vector: torch.Tensor = ctx.saved_tensors[0]
+        group: torch.Tensor = ctx.saved_tensors[1]
+        if not hasattr(ctx, "jacobians"):
+            ctx.jacobians: torch.Tensor = _j_exp_map_impl(tangent_vector)
+        jacs = ctx.jacobians
+        dR = group.transpose(1, 2) @ grad_output
+        grad_input = jacs.transpose(1, 2) @ torch.stack(
+            (
+                dR[:, 2, 1] - dR[:, 1, 2],
+                dR[:, 0, 2] - dR[:, 2, 0],
+                dR[:, 1, 0] - dR[:, 0, 1],
+            ),
+            dim=1,
+        ).view(-1, 3, 1)
+        return grad_input.view(-1, 3)
+
+
+_module = get_module(__name__)
+
+exp_map, jexp_map = LieGroup.ExpMapFactory(_module)
```

Comment:
If you pass `ExpMap` and `_j_exp_map_impl` as arguments, no need to use this `get_module`.